### PR TITLE
API: Remove zero names from dtype aliases

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @rgommers @ngoldbaum @seberg,

This PR removes 7 names from the `np.dtype(<str_alias>)` search:
```
int0
uint0
void0
object0
str0
bytes0
bool8
``` 
as these have been deprecated in https://github.com/numpy/numpy/pull/22607 in 1.24 release. Their `np.*` types counterparts have been already removed from the main namespace.

For informational purposes I added `if` check in the C code that explicitly verifies if one of these names is provided to display a custom error message. I'm not sure how much overhead such additional check will have on `np.dtype` but it's only evaluated when the execution is going to fail anyway with `data type *** not understood`.

### CODE REVIEW

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1845,6 +1845,14 @@ _convert_from_str(PyObject *obj, int align)
             if (PyErr_Occurred()) {
                 return NULL;
             }
+            if (
+                strcmp(type, "int0") == 0 || strcmp(type, "uint0") == 0 ||
+                strcmp(type, "void0") == 0 || strcmp(type, "object0") == 0 ||
+                strcmp(type, "str0") == 0 || strcmp(type, "bytes0") == 0 ||
+                strcmp(type, "bool8") == 0
+            ) {
+                goto expired_alias_fail;
```
Comment (seberg) [inline numpy/core/src/multiarray/descriptor.c:1854]: The goto isn't necessary for a single place (at least without cleanup, cleanup could make it worthwhile).

**Review by seberg — APPROVED** [review]
Looks fine to me, not sure I am use dto black style in C code yet, but doesn't matter...

`int0` going to `intp` (IIRC) is rather confusing, but I don't understand what the point of these ever were and this was already pretty much deprecated.

Anyway, happy with this, unless someone points out that this is used a lot downstream and I really doubt that here.

File: numpy/core/src/multiarray/descriptor.c (hunk 1 above)
Comment (charris) [reply numpy/core/src/multiarray/descriptor.c:1854]: @mtsokol Might as well clean this up a bit so we can merge it.

Comment (mtsokol) [reply numpy/core/src/multiarray/descriptor.c]: I removed the redundant `goto` statement. Let me know if I should clean the code further!

File: numpy/core/_type_aliases.py [hunk 2]
```diff
@@ -224,9 +222,7 @@ def _set_array_types():
 
 # Add additional strings to the sctypeDict
 _toadd = ['int', ('float', 'double'), ('complex', 'cdouble'), 
-          'bool', 'object',
-          'str', 'bytes', ('a', 'bytes_'),
-          ('int0', 'intp'), ('uint0', 'uintp')]
+          'bool', 'object', 'str', 'bytes', ('a', 'bytes_')]
```
Comment (rgommers) [inline numpy/core/_type_aliases.py:225]: `np.dtype('a')` also seems ripe for removal, never seen it before and looks very odd. Does anyone know where that came from?

**Review by rgommers — APPROVED** [review]
LGTM, let's give this a go - thanks Mateusz & reviewers!

File: numpy/core/_type_aliases.py (hunk 2 above)
Comment (mtsokol) [reply numpy/core/_type_aliases.py:225]: I've actually noticed `np.dtype('a')` but I couldn't figure out its origin and purpose 😄 - therefore I decided to keep it.

Comment (ngoldbaum) [reply numpy/core/_type_aliases.py:225]: Looks like this goes back to https://github.com/numpy/numpy/commit/9ac3b0e5c2e181873221ff0160b882772ea9d435. I have no idea why Travis added it and I don't know whether issues from the old trac issue tracker are still visible somewhere. Github search only finds 12 code usages of `"np.dtype('a')"`. I say remove it.

Comment (seberg) [reply numpy/core/_type_aliases.py:225]: Presumably it is meant to mean "ascii"... if almost nobody uses, it shouldn't matter, unlik `"c"` it dosn't eve have a special meaning.

**Comment by EpicWink** [comment]
Am I to understand that these will be removed in NumPy v2.0? Or will it happen in a v1.x release?

**Comment by mtsokol** [comment]
> Am I to understand that these will be removed in NumPy v2.0? Or will it happen in a v1.x release?

These changes are planned for NumPy 2.0 release. [Here](https://github.com/numpy/numpy/issues?q=label%3A%22Numpy+2.0+API+Changes%22+is%3Aclosed) you can find all PRs with 2.0 changes label.

**Pull request merged**